[Pipeline] Add new pipeline for ParaDiGMS -- parallel sampling of diffusion models

AndyShih12 opened the pull request:
This pull request implements the paper Parallel Sampling of Diffusion Models: https://arxiv.org/abs/2305.16317
Based on the repository: https://github.com/AndyShih12/paradigms

Example of use:
```python
import torch
from diffusers import DDPMScheduler
from diffusers import StableDiffusionParadigmsPipeline

scheduler = DDPMScheduler.from_pretrained("runwayml/stable-diffusion-v1-5", subfolder="scheduler")

pipe = StableDiffusionParadigmsPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", scheduler=scheduler, torch_dtype=torch.float16)
pipe = pipe.to("cuda")

ngpu, batch_per_device = torch.cuda.device_count(), 5
pipe.wrapped_unet = torch.nn.DataParallel(pipe.unet, device_ids=[d for d in range(ngpu)])

prompt = "a photo of an astronaut riding a horse on mars"
image = pipe(prompt, parallel=ngpu * batch_per_device, num_inference_steps=1000).images[0]
```

alexblattner commented:
@AndyShih12 can this be used with other pipelines too to make them concurrent?

AndyShih12 commented:
Yes paradigms should be compatible with most pipelines, just by changing the denoising loop to the parallel denoising loop. But each pipeline is in a different file so we would have to make a parallel version of each pipeline separately.

AndyShih12 commented:
requesting feedback @patrickvonplaten @sayakpaul

alexblattner commented:
@patrickvonplaten @sayakpaul I think that there's an issue with the way pipelines are made. This pipeline will only work for txt2img only despite needing a specific part modified only. I think there should be a better way to mix multiple pipeline than making more and more pipeline. I am unsure what the best approach would be, but I think this would save a lot of work in the long run.

sayakpaul commented:
> @patrickvonplaten @sayakpaul I think that there's an issue with the way pipelines are made. This pipeline will only work for txt2img only despite needing a specific part modified only. I think there should be a better way to mix multiple pipeline than making more and more pipeline. I am unsure what the best approach would be, but I think this would save a lot of work in the long run.

I understand your point of view. But as stated in our [documentation](https://huggingface.co/docs/diffusers/main/en/index), we want to be flexible and readable with our codebase. The way we achieve that is by exposing the pipelines in a way that is as self-contained as possible, taking inspiration from `transformers`. Ccing @patrickvonplaten to share more on this if needed.

sayakpaul commented:
@AndyShih12 thanks so much for your PR. 

Could you also give us some comparisons on how parallel sampling improves the efficiency of a standard say, `StableDiffusionPipeline`?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_3716). All of your documentation changes will be reflected on that endpoint.

AndyShih12 commented:
@sayakpaul thank you!

Sure, parallel sampling on 8 GPUs can give a 3.1x speedup over StableDiffusionPipeline for 1000-step DDPM, and a 1.8x speedup for 200-step DDIM. There are some more details and comparisons in the [paper](https://arxiv.org/abs/2305.16317).

<img width="1440" alt="results_table" src="https://github.com/huggingface/diffusers/assets/11932455/a18f18eb-71c3-43da-b32b-86f6ede8096f">

Here is what we can expect the speedup to be when using fewer GPUs, on 1000-step DDPM.

![paraddpm](https://github.com/huggingface/diffusers/assets/11932455/7a343473-5f13-426a-872c-78053542a433)


Here is a script to compare running time. As we can see above, it's important run with multiple GPUs.
```python
import torch
from diffusers import DDPMScheduler, DDIMScheduler
from diffusers import StableDiffusionParadigmsPipeline, StableDiffusionPipeline

scheduler = DDPMScheduler.from_pretrained("runwayml/stable-diffusion-v1-5", subfolder="scheduler")
prompt = "a photo of an astronaut riding a horse on mars"

sequential_pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", scheduler=scheduler, torch_dtype=torch.float16)
sequential_pipe = sequential_pipe.to("cuda")

ngpu = torch.cuda.device_count()
parallel_pipe = StableDiffusionParadigmsPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", scheduler=scheduler, torch_dtype=torch.float16)
parallel_pipe = parallel_pipe.to("cuda")
parallel_pipe.wrapped_unet = torch.nn.DataParallel(parallel_pipe.unet, device_ids=[d for d in range(ngpu)])

num_inference_steps, batch_per_device = 1000, 5
# warmup
_ = sequential_pipe(prompt, num_inference_steps=10).images[0]
_ = parallel_pipe(prompt, parallel=ngpu * batch_per_device, num_inference_steps=10).images[0]
# run
image_sequential = sequential_pipe(prompt, num_inference_steps=num_inference_steps).images[0]
image_parallel = parallel_pipe(prompt, parallel=ngpu * batch_per_device, num_inference_steps=num_inference_steps).images[0]
```

sayakpaul commented:
Thanks for providing the info!

> Sure, parallel sampling on 8 GPUs can give a 3.1x speedup over StableDiffusionPipeline for 1000-step DDPM, and a 1.8x speedup for 200-step DDIM. 

Do you mean that if we use the pipeline **_without_** parallel sampling on `StableDiffusionPipeline` and run inference with 8 GPUs for 1000-step DDPM, it's 3.1x slower? That's quite some speedup!

Additionally, I wanted to make you aware of our distributed inference support in case you aren't: https://huggingface.co/docs/diffusers/main/en/training/distributed_inference. It would be good to have a direct comparison of the timings using the setup used in our docs as well. 

Nonetheless, I am reviewing your PR now.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
```suggestion
# Copyright 2023 ParaDiGMS authors and The HuggingFace Team. All rights reserved.
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+
+class StableDiffusionParadigmsPipeline(DiffusionPipeline, TextualInversionLoaderMixin, LoraLoaderMixin, FromCkptMixin):
+    r"""
+    Parallelized version of StableDiffusionPipeline, based on the paper https://arxiv.org/abs/2305.16317
```
I think it's better to clarify what kind of parallelism we're achieving here, i.e., data-parallel, model-parallel, etc.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.register_to_config(requires_safety_checker=requires_safety_checker)
+
+        self.wrapped_unet = self.unet
```
Since we don't have an attribute in the rest of the pipelines called `wrapped_unet`, contributors might wonder what this is doing here. 

Maybe adding a short comment would be helpful?

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        self.register_to_config(requires_safety_checker=requires_safety_checker)
+
+        self.wrapped_unet = self.unet
+
```
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.enable_vae_slicing
```

Since we're using a copied version here.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        steps. This is useful to save some memory and allow larger batch sizes.
+        """
+        self.vae.enable_slicing()
+
```
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.disable_vae_slicing
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        computing decoding in one step.
+        """
+        self.vae.disable_slicing()
+
```
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.enable_vae_tiling
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        several steps. This is useful to save a large amount of memory and to allow the processing of larger images.
+        """
+        self.vae.enable_tiling()
+
```
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.disable_vae_tiling
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        computing decoding in one step.
+        """
+        self.vae.disable_tiling()
+
```
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.enable_sequential_cpu_offload
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+
+        if self.safety_checker is not None:
+            cpu_offload(self.safety_checker, execution_device=device, offload_buffers=True)
+
```
```suggestion
# Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.enable_model_cpu_offload
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
+    @property
```
```suggestion
    @property
    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._execution_device
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
+
```
```suggestion
    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._encode_prompt
```

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+            prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds])
+
+        return prompt_embeds
+
```
To add: `# Copied from ...` comment.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+                images=image, clip_input=safety_checker_input.pixel_values.to(dtype)
+            )
+        return image, has_nsfw_concept
+
```
To add: `# Copied from ...` comment.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        # we always cast to float32 as this does not cause significant overhead and is compatible with bfloat16
+        image = image.cpu().permute(0, 2, 3, 1).float().numpy()
+        return image
+
```
To add: `# Copied from ...` comment.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        if accepts_generator:
+            extra_step_kwargs["generator"] = generator
+        return extra_step_kwargs
+
```
To add: `# Copied from ...` comment.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        return latents
+
+    def _cumsum(self, input, dim, debug=False):
+        if debug:
+            # cumsum_cuda_kernel does not have a deterministic implementation
+            # so perform cumsum on cpu for debugging purposes
+            return torch.cumsum(input.cpu().float(), dim=dim).to(input.device)
```
Super nice!

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        num_inference_steps: int = 50,
+        parallel: int = 10,
```
I think we're missing docstrings for this argument. In general, let's ensure we have docstrings for all the positional arguments for this function.

sayakpaul commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,847 @@
+        end_idx = parallel
+        latents_time_evolution_buffer = torch.stack([latents] * (len(scheduler.timesteps) + 1))
+
+        noise_array = torch.zeros_like(latents_time_evolution_buffer)
+        for j in range(len(scheduler.timesteps)):
+            base_noise = randn_tensor(
+                shape=latents.shape, generator=generator, device=latents.device, dtype=prompt_embeds.dtype
+            )
+            noise = (self.scheduler._get_variance(scheduler.timesteps[j]) ** 0.5) * base_noise
+            noise_array[j] = noise.clone()
+        inverse_variance_norm = 1.0 / torch.tensor(
+            [scheduler._get_variance(scheduler.timesteps[j]) for j in range(len(scheduler.timesteps))] + [0]
+        ).to(noise_array.device)
+        latent_dim = noise_array[0, 0].numel()
+        inverse_variance_norm = inverse_variance_norm[:, None] / latent_dim
+
+        scaled_tolerance = tolerance**2
```
Could we have a comment for this code block on what it is doing?

sayakpaul commented on src/diffusers/schedulers/scheduling_ddim.py:
```diff
@@ -444,6 +459,124 @@ def step(
+        return_dict: bool = True,
+    ) -> Union[DDIMSchedulerOutput, Tuple]:
+        """
+        Predict the sample at the previous timestep by reversing the SDE. Core function to propagate the diffusion
+        process from the learned model outputs (most often the predicted noise).
```
IIUC it's the batched implementation, if so, I'd clarify that in the docstrings.

sayakpaul commented on src/diffusers/schedulers/scheduling_ddpm.py:
```diff
@@ -405,6 +406,71 @@ def step(
+        sample: torch.FloatTensor,
+    ) -> Union[DDPMSchedulerOutput, Tuple]:
+        """
+        Predict the previous timestep of multiple samples in batch. Does not add noise.
```
Same as https://github.com/huggingface/diffusers/pull/3716/files#r1227448253. But also, I think we're missing docstrings.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,219 @@
+    def test_inference_batch_single_identical(self):
+        super().test_inference_batch_single_identical(batch_size=2, expected_max_diff=3e-3)
+
+    def test_stable_diffusion_paradigms_negative_prompt(self):
```
Maybe also add a test for DDPM since it's supported?

Also, IIUC, to make this supported by the rest of the schedulers, we need to implement the `batch_step_no_noise()` inside of the scheduler class, correct?

AndyShih12 replied:
yes, that is correct.

sayakpaul commented on tests/pipelines/stable_diffusion/test_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,219 @@
+
+        assert image.shape == (1, 512, 512, 3)
+
+        expected_slice = np.array(
+            [
+                0.96222240,
+                0.96028924,
+                0.97488534,
+                0.95912700,
+                0.96303370,
+                0.96918930,
+                0.96610457,
+                0.96318950,
+                0.97419780,
+            ]
+        )
```
Which env was this tested on? PT 2.0?

sayakpaul replied:
Also, I think we can reduce the LoC here like so:

https://github.com/huggingface/diffusers/blob/ce5504934ac484fca39a1a5434ecfae09eabdf41/tests/pipelines/stable_diffusion/test_stable_diffusion_pix2pix_zero.py#L395

AndyShih12 replied:
yes, PT 2.0

sayakpaul reviewed (commented):
This is a very nice PR. Thanks much!

My comments are on the doc side of things. Generally:

* We also need to have a dedicated doc page for `ParaDiGMS` like the other pipelines we have ([reference](https://github.com/huggingface/diffusers/tree/main/docs/source/en/api/pipelines)). I'd envision the documentation to have the benchmarks on the timing and the requirements for ParaDiGMS to perform well. 
* The pipeline code looks mostly good to me, however, there were times the code wasn't very readable. So, I would suggest grouping logically coherent pieces of code blocks together and accompanying them with comments. This is important because many of our users directly refer to the codebase to understand things. 

Let me know if anything is unclear or anything I can do to help.

AndyShih12 commented:
Thank you so much @sayakpaul for the detailed review, glad to hear that it is a nice PR! 

I've incorporated new changes based on all the feedback, and added a dedicated doc page as well. Let me know if any additional edits are needed 😄

AndyShih12 commented:
Addressing your earlier question:

> Do you mean that if we use the pipeline **_without_** parallel sampling on `StableDiffusionPipeline` and run inference with 8 GPUs for 1000-step DDPM, it's 3.1x slower? That's quite some speedup!

Yes, this is correct.

sayakpaul commented:
@AndyShih12 could you please run `make style && make quality` from your end and push the changes? Instructions are available here:
https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md

sayakpaul commented on docs/source/en/api/pipelines/paradigms.mdx:
```diff
@@ -0,0 +1,69 @@
+may not fit in memory, and lower batch size gives less parallelism. 
+- For tolerance, using a higher tolerance may get better speedups but can risk sample quality degradation. 
+If there is quality degradation with the default tolerance, then use a lower tolerance (e.g. 0.001).
+</Tip>
```
Maybe we can also include information on the expected speedup here? This helps the users gauge things better. WDYT?

AndyShih12 replied:
ok added!

sayakpaul commented:
@AndyShih12 this PR looks good to me! Maybe we'd want to also add tests for the changes introduced in the DDIM and DDPM scheduler? 

Additionally, once https://github.com/huggingface/diffusers/pull/3716#issuecomment-1590621021 is addressed, I think we can prep for merging!

AndyShih12 commented:
Thanks for the feedback! I added test cases for batch_step_no_noise for DDPM and DDIM scheduler, also included an entry to the doc about the distributed_inference support, and a comment about expected speedup. Let me know if there's anything else!

sayakpaul commented on docs/source/en/api/pipelines/paradigms.mdx:
```diff
@@ -0,0 +1,81 @@
+</Tip>
+
+<Tip>
+Diffusers also offers distributed inference support for generating multiple prompts in parallel on multiple GPUs.
```
```suggestion
Diffusers also offers distributed inference support for generating multiple prompts in parallel on multiple GPUs. Check out the docs [here](https://huggingface.co/docs/diffusers/main/en/training/distributed_inference).
```

sayakpaul commented on docs/source/en/api/pipelines/paradigms.mdx:
```diff
@@ -0,0 +1,81 @@
+
+<Tip>
+Diffusers also offers distributed inference support for generating multiple prompts in parallel on multiple GPUs.
+https://huggingface.co/docs/diffusers/main/en/training/distributed_inference
```
```suggestion
```

sayakpaul reviewed (approved):
Thanks for iterating so much! PR looks good to me. 

I would also like a review from @patrickvonplaten.

sayakpaul commented:
@AndyShih12 for this [failure](https://github.com/huggingface/diffusers/actions/runs/5270332687/jobs/9538132998?pr=3716), from your "quality" virtual Python environment, you can run `make fix-copies`.

AndyShih12 commented:
Got it, thanks for the edits!

sayakpaul commented:
Failing tests are unrelated.

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,896 @@
+                " checker. If you do not want to use the safety checker, you can pass `'safety_checker=None'` instead."
+            )
+
+        is_unet_version_less_0_9_0 = hasattr(unet.config, "_diffusers_version") and version.parse(
+            version.parse(unet.config._diffusers_version).base_version
+        ) < version.parse("0.9.0.dev0")
+        is_unet_sample_size_less_64 = hasattr(unet.config, "sample_size") and unet.config.sample_size < 64
+        if is_unet_version_less_0_9_0 and is_unet_sample_size_less_64:
+            deprecation_message = (
+                "The configuration file of the unet has set the default `sample_size` to smaller than"
+                " 64 which seems highly unlikely. If your checkpoint is a fine-tuned version of any of the"
+                " following: \n- CompVis/stable-diffusion-v1-4 \n- CompVis/stable-diffusion-v1-3 \n-"
+                " CompVis/stable-diffusion-v1-2 \n- CompVis/stable-diffusion-v1-1 \n- runwayml/stable-diffusion-v1-5"
+                " \n- runwayml/stable-diffusion-inpainting \n you should change 'sample_size' to 64 in the"
+                " configuration file. Please make sure to update the config accordingly as leaving `sample_size=32`"
+                " in the config might lead to incorrect results in future versions. If you have downloaded this"
+                " checkpoint from the Hugging Face Hub, it would be very nice if you could open a Pull request for"
+                " the `unet/config.json` file"
+            )
+            deprecate("sample_size<64", "1.0.0", deprecation_message, standard_warn=False)
+            new_config = dict(unet.config)
+            new_config["sample_size"] = 64
+            unet._internal_dict = FrozenDict(new_config)
+
```
```suggestion
```

patrickvonplaten commented on src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_paradigms.py:
```diff
@@ -0,0 +1,896 @@
+            )
+        return image, has_nsfw_concept
+
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.decode_latents
+    def decode_latents(self, latents):
+        warnings.warn(
+            "The decode_latents method is deprecated and will be removed in a future version. Please"
+            " use VaeImageProcessor instead",
+            FutureWarning,
+        )
+        latents = 1 / self.vae.config.scaling_factor * latents
+        image = self.vae.decode(latents, return_dict=False)[0]
+        image = (image / 2 + 0.5).clamp(0, 1)
+        # we always cast to float32 as this does not cause significant overhead and is compatible with bfloat16
+        image = image.cpu().permute(0, 2, 3, 1).float().numpy()
+        return image
+
```
```suggestion
```
Function is deprecated

patrickvonplaten commented on src/diffusers/schedulers/scheduling_ddim.py:
```diff
@@ -170,6 +170,7 @@ class DDIMScheduler(SchedulerMixin, ConfigMixin):
 
     _compatibles = [e.name for e in KarrasDiffusionSchedulers]
```
Could we create a new scheduler class here, `DDIMParallelScheduler`?

I'd like to keep the DDIM scheduler simple if possible :-)

AndyShih12 replied:
ok done!

patrickvonplaten commented on src/diffusers/schedulers/scheduling_ddpm.py:
```diff
@@ -118,6 +118,7 @@ class DDPMScheduler(SchedulerMixin, ConfigMixin):
 
     _compatibles = [e.name for e in KarrasDiffusionSchedulers]
     order = 1
+    _is_ode_scheduler = False
```
Could we create a new scheduler class here, `DDPMParallelScheduler`?

I'd like to keep the DDPM scheduler simple if possible :-)

AndyShih12 replied:
ok done!

AndyShih12 commented:
Ok, I've refactored the schedulers so that DDPMParallelScheduler and DDIMParallelScheduler are now separate classes! @patrickvonplaten

sayakpaul commented on src/diffusers/schedulers/scheduling_ddim.py:
```diff
@@ -249,6 +252,17 @@ def _get_variance(self, timestep, prev_timestep):
 
         return variance
 
+    def _batch_get_variance(self, t, prev_t):
```
Do we need these modifications when we have them already in the parallel versions?

sayakpaul commented on tests/schedulers/test_schedulers.py:
```diff
@@ -238,6 +238,12 @@ def get_scheduler_config(self):
 
     def dummy_model(self):
         def model(sample, t, *args):
+            # if t is a tensor, match the number of dimensions of sample
+            if isinstance(t, torch.Tensor):
+                num_dims = len(sample.shape)
+                # pad t with 1s to match num_dims
+                t = t.reshape(-1, *(1,) * (num_dims - 1))
+
```
Could I get a brief explanation on why was this needed?

patrickvonplaten replied:
Also curious here

AndyShih12 replied:
Typically timestep `t` is a scalar, and `sample` has shape e.g. (batch, 3, 8, 8).
For parallel sampling, `t` has shape (batch,) and `sample` has shape (batch, 3, 8, 8).

When running net.forward(sample, t, ...) in with the typical unet, this runs fine.

sayakpaul commented on tests/schedulers/test_scheduler_ddpm_parallel.py:
```diff
@@ -0,0 +1,212 @@
+            # 2. predict previous mean of sample x_t-1
+            pred_prev_sample = scheduler.step(residual, t, sample, generator=generator).prev_sample
+
+            # if t > 0:
+            #     noise = self.dummy_sample_deter
+            #     variance = scheduler.get_variance(t) ** (0.5) * noise
+            #
+            # sample = pred_prev_sample + variance
```
We can remove this.

sayakpaul commented on tests/schedulers/test_scheduler_ddpm_parallel.py:
```diff
@@ -0,0 +1,212 @@
+            # 2. predict previous mean of sample x_t-1
+            pred_prev_sample = scheduler.step(residual, t, sample, generator=generator).prev_sample
+
+            # if t > 0:
+            #     noise = self.dummy_sample_deter
+            #     variance = scheduler.get_variance(t) ** (0.5) * noise
+            #
+            # sample = pred_prev_sample + variance
```
Same.

sayakpaul commented on src/diffusers/schedulers/scheduling_ddim_parallel.py:
```diff
@@ -0,0 +1,627 @@
+# Copyright 2023 Stanford University Team and The HuggingFace Team. All rights reserved.
```
I think you should also retain the credits here since your team introduced the changes proposed here. 

@patrickvonplaten WDYT?

patrickvonplaten replied:
Yeah guess that would be nicer

sayakpaul commented on tests/schedulers/test_scheduler_ddpm_parallel.py:
```diff
@@ -0,0 +1,212 @@
+import torch
+
+from diffusers import DDPMParallelScheduler
+
+from .test_schedulers import SchedulerCommonTest
```
Licensing missing.

sayakpaul commented on tests/schedulers/test_scheduler_ddim_parallel.py:
```diff
@@ -0,0 +1,174 @@
+import torch
+
+from diffusers import DDIMParallelScheduler
+
+from .test_schedulers import SchedulerCommonTest
```
Licensing missing.

sayakpaul commented on src/diffusers/schedulers/scheduling_ddpm_parallel.py:
```diff
@@ -0,0 +1,558 @@
+# Copyright 2023 UC Berkeley Team and The HuggingFace Team. All rights reserved.
```
Same.

sayakpaul commented on src/diffusers/schedulers/scheduling_ddim_parallel.py:
```diff
@@ -0,0 +1,627 @@
+        t = t.view(-1, *([1] * (model_output.ndim - 1)))
+        prev_t = prev_t.view(-1, *([1] * (model_output.ndim - 1)))
+
+        # if model_output.shape[1] == sample.shape[1] * 2 and self.variance_type in ["learned", "learned_range"]:
+        #     model_output, predicted_variance = torch.split(model_output, sample.shape[1], dim=1)
+        # else:
+        #     predicted_variance = None
```
I guess we can remove this.

patrickvonplaten commented on src/diffusers/schedulers/scheduling_ddim.py:
```diff
@@ -239,7 +239,10 @@ def scale_model_input(self, sample: torch.FloatTensor, timestep: Optional[int] =
         return sample
 
-    def _get_variance(self, timestep, prev_timestep):
+    def _get_variance(self, timestep, prev_timestep=None):
```
Could we indeed try to remove all modifications to this file?

patrickvonplaten reviewed (commented):
Nice, think we're close to getting this one in

AndyShih12 commented:
Awesome, I made the new changes so that scheduling_ddim.py is now untouched! Let me know if there is anything else!

patrickvonplaten commented:
It looks like a couple of scheduler tests are now failing:
```
FAILED tests/schedulers/test_scheduler_euler_ancestral.py::EulerAncestralDiscreteSchedulerTest::test_full_loop_device - assert 2.4515058968472374 < 0.01
 +  where 2.4515058968472374 = abs((154.77070589684723 - 152.3192))
 +    where 154.77070589684723 = <built-in method item of Tensor object at 0x7f19eaf3e9a0>()
 +      where <built-in method item of Tensor object at 0x7f19eaf3e9a0> = tensor(154.7707, dtype=torch.float64).item
FAILED tests/schedulers/test_scheduler_euler_ancestral.py::EulerAncestralDiscreteSchedulerTest::test_full_loop_no_noise - assert 2.4515058968472374 < 0.01
 +  where 2.4515058968472374 = abs((154.77070589684723 - 152.3192))
 +    where 154.77070589684723 = <built-in method item of Tensor object at 0x7f19eb100b80>()
 +      where <built-in method item of Tensor object at 0x7f19eb100b80> = tensor(154.7707, dtype=torch.float64).item
FAILED tests/schedulers/test_scheduler_euler_ancestral.py::EulerAncestralDiscreteSchedulerTest::test_full_loop_with_v_prediction - assert 4.660559621037578 < 0.01
 +  where 4.660559621037578 = abs((103.78334037896242 - 108.4439))
 +    where 103.78334037896242 = <built-in method item of Tensor object at 0x7f19dfd93bd0>()
 +      where <built-in method item of Tensor object at 0x7f19dfd93bd0> = tensor(103.7833, dtype=torch.float64).item
FAILED tests/schedulers/test_scheduler_kdpm2_ancestral.py::KDPM2AncestralDiscreteSchedulerTest::test_full_loop_device - assert 165.99396548951336 < 0.1
 +  where 165.99396548951336 = abs((14015.375765489513 - 13849.3818))
 +    where 14015.375765489513 = <built-in method item of Tensor object at 0x7f19ead8f360>()
 +      where <built-in method item of Tensor object at 0x7f19ead8f360> = tensor(14015.3758, dtype=torch.float64).item
FAILED tests/schedulers/test_scheduler_kdpm2_ancestral.py::KDPM2AncestralDiscreteSchedulerTest::test_full_loop_no_noise - assert 165.98806548951325 < 0.01
 +  where 165.98806548951325 = abs((14015.375765489513 - 13849.3877))
 +    where 14015.375765489513 = <built-in method item of Tensor object at 0x7f19eb0ac310>()
 +      where <built-in method item of Tensor object at 0x7f19eb0ac310> = tensor(14015.3758, dtype=torch.float64).item
FAILED tests/schedulers/test_scheduler_kdpm2_ancestral.py::KDPM2AncestralDiscreteSchedulerTest::test_full_loop_with_v_prediction - assert 4.016957222634858 < 0.01
 +  where 4.016957222634858 = abs((333.0139572226349 - 328.997))
 +    where 333.0139572226349 = <built-in method item of Tensor object at 0x7f19dfa89e50>()
 +      where <built-in method item of Tensor object at 0x7f19dfa89e50> = tensor(333.0140, dtype=torch.float64).item
```

Could we try to fix those before merging? :-)

sayakpaul commented:
I ran the failing test cases (`test_scheduler_kdpm2_ancestral.py` and `test_scheduler_euler_ancestral.py`) from the main branch of `diffusers`. They passed successfully whereas when I ran them from @AndyShih12's branch they didn't. 

@AndyShih12, let's maybe try to fix the failing tests before merging :)

AndyShih12 commented:
Oh thank you for the catch, I had assumed the test issue was due to the main branch. Sorry about not testing it thoroughly.

Indeed I fixed it by properly casting the type/device of the new tensor in the dummy model.

sayakpaul commented:
The failing test seems to be unrelated. Let me update the branch once and rerun the workflow.

sayakpaul commented:
Thanks for your great contribution!

AndyShih12 commented:
thank you both for all the help!
